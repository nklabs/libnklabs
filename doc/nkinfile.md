# Input file abstraction

## Files

[nkinfile.h](../inc/nkinfile.h),
[nkinfile.c](../src/nkinfile.c)

## Description

```c
nkinfile_t *nkinfile_open(
    nkinfile_t *f,
    size_t (*block_read)(
        void *block_read_ptr,
        size_t offset,
        unsigned char *buffer,
	size_t size
    ),
    void *block_read_ptr,
    size_t block_size,
    unsigned char *buffer
);

nkinfile_t *nkinfile_open_mem(nkinfile_t *f, unsigned char *mem, size_t size);

nkinfile_t *nkinfile_open_string(nkinfile_t *f, char *s);

size_t nk_ftell(nkinfile_t *f);

int nk_feof(nkinfile_t *f);

int nk_fseek(nkinfile_t *f, size_t pos);

int nk_fseek_rel(nkinfile_t *f, size_t rel_pos);

int nk_fpeek(nkinfile_t *f);

int nk_fpeek_rel(nkinfile_t *f, size_t rel_pos);

int nk_fpeek_abs(nkinfile_t *f, size_t pos);

int nk_fnext(nkinfile_t *f);
int nk_fnext_fast(nkinfile_t *f);

int nk_fgetc(nkinfile_t *f);
int nk_fgetc_fast(nkinfile_t *f);

int nk_infile_print(nkinfile_t *f);

int nk_fcopy(nkoutfile_t *g, nkinfile_t *f);
```

This is a thin wrapper that provides fast, low overhead file-like random
read access primitives to an underlying memory area, C-string or block
device.  It is used to direct input to __nk_fscanf__.  __nk_fscanf__
requires random access since it restores the byte offset to the original
location (before __nk_fscan__ was called) when it fails to parse.

An nkinfile_t may be opened in the following ways:

__nkinfile_open_string__ opens a NUL-terminated C string as a file.  It
calls __strlen__ to find the length of the string.

__nkinfile_open_mem__ opens a memory area of a given length as a file.

__nkinfile_open__ opens some kind of block device as a file.  __block_read__
is a function you provide which should read a block from the device at a
given __offset__ (which will be a multiple of __block_size__) to the
specified memory buffer.  __block_read_ptr__ is a void pointer passed
unchanged as the first argument of __block_read__ whenever it is called. 
__block_size__ specifies the device's block size, which is the number of
bytes that should be transferred by a single call to __block_read__. 
__buffer__ is a memory buffer that you provide that must be large enough
for __block_size__ bytes.

Note that __block_read__ is immediately called by __nkinfile_open__ to load
the first block of the file into the buffer.

__block_read__ must return the number of bytes read.  This should always be
equal to __block_size__ except for when the last block of the file (or one
past the last block of the file for the case where the last block contains
__block_size__ bytes) is read.  This is how the end of file is indicated to
the nkinfile wrapper.

Note that __block_read__ may be called with any valid position (any position
which is a multiple of the block size) in any order.  Note that even if the
last block has been read, blocks from other offsets may still be
subsequently requested.

__nkinfile_open__, __nkinfile_open_mem__ and __nkinfile_open_string__ all
return the __nkinfile_t__ pointer.

Note that there is no file close operation.  The __nkinfile_t__ does not
allocate any resources that need to be freed or released.  Releasing the
buffer (if necessary) is up to you.

__nk_feof__ returns true if the nkinfile_t byte offset is at the end of the
file.  __nk_feof__ is implemented as a macro.

__nk_ftell__ returns the byte offset of the nkinfile_t.  __nk_ftell__ is
implemented as a macro.

__nk_fseek__ repositions the nkinfile_t to the requested byte offset.  The
byte offset must be in the range 0 to the size of the file inclusive.  If
the file size is 4 bytes, then the possible valid byte offsets are 0, 1, 2,
3 and 4.  __nk_fseek__ is implemented as a macro.

__nk_fseek_rel__ is like __nk_fseek__, except that it performs a relative
seek- that is, it repositions the nkinfile_t to nk_ftell(f) + rel_pos. 
nk_ftell(f) + rel_pos must be in the range 0 to the size of the file
inclusive.

__nk_fpeek__ reads a single unsigned byte from the input file, without
changing the current byte offset.  If the byte offset is at the end of the
file, __nk_fpeek__ returns -1.  __nk_fpeek__ is implemented as a macro.

__nk_fseek_rel__ is like __nk_fpeek__ except that it reads from nk_ftell(f) + __rel_pos__.

__nk_fseek_abs__ is like __nk_fpeek__ except that it reads the byte at byte
offset __pos__.

__nk_fnext__ advances the byte offset by 1, and returns the byte at that
location.  It returns -1 if the new byte offset is at the end of the file. 
Note that if the byte offset was already at the end, __nk_fnext__ does not
change the offset and it returns -1.  __nk_fnext_fast__ may be used if it is
known that the byte offset is not already at the end of the file (via a
previous call to __nk_feof__ or __nk_fpeek__).  __nk_fnext__ and
__nk_fnext_fast__ are implemented as macros.

__nk_fgetc__ reads the byte at the current byte offset, and then advances
the offset by 1.  Note that if the offset was at the end of the file,
__nk_fgetc__ returns -1 and does not change the offset.  __nk_fgetc_fast__
may be used if it is known that the byte offset is not already at the end
(via a previous call to __nk_feof__ or __nk_fpeek__).  __nk_fgetc__ and
__nk_fgetc_fast__ are implemented as macros.

__nk_fcopy__ copies the remainder of the nkinfile_t to the specified
nkoutfile_t.

__nk_infile_print__ is the same as __nk_fcopy__ except that it copies from
the specified __nkinfile_t__ to __nkstdout__.  It prints the remainder of the
nkinfile_t.
