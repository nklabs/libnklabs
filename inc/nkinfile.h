// File access abstraction

//  Access a string as a file
//  Access a memory block as a file
//  Access a block device as a file

// Supports random access through tell/seek

// This was written for nkscan and nkdbase, so that we can parse a database
// located in SPI-flash without first having to copy the string to a large
// RAM buffer.  Instead all of the parsing functions read the input as if
// it was a file, where the device level access is provided via block reads
// to a small buffer.

// Copyright 2021 NK Labs, LLC

// Permission is hereby granted, free of charge, to any person obtaining a
// copy of this software and associated documentation files (the
// "Software"), to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to permit
// persons to whom the Software is furnished to do so, subject to the
// following conditions:

// The above copyright notice and this permission notice shall be included
// in all copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. 
// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT
// OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR
// THE USE OR OTHER DEALINGS IN THE SOFTWARE.

#ifndef _nkinfile
#define _nkinfile

#include <stdlib.h> // For ssize_t

#include "nkoutfile.h"

// Nkfile pointer

typedef struct nkinfile nkinfile_t;

struct nkinfile
{
    int tmp; // Temporary character
    size_t tmp_pos; // Temporary position for nkfpeek_rel and nkfpeek_abs

    // Information about current buffer
    unsigned char *buffer; // Buffer we read into
    const unsigned char *ptr; // Current pointer into current buffer
    const unsigned char *start; // Start of current buffer (usually same as 'buffer', but const)
    const unsigned char *end; // End of current buffer
    size_t len; // Number of bytes available in current buffer
    size_t start_offset; // File offset of first byte of current buffer

    // For abstracting some kind of fixed-length block-based storage system into a file
    void *block_read_ptr; // Pointer to pass to block_read
    size_t (*block_read)(void *block_read_ptr, size_t pos, unsigned char *buffer, size_t block_size);
    size_t block_size; // Block size, also buffer must be this size
};

// Return current file position
// Return 0..file_size.  If it returns file_size, it means we are at end of file.
#define nk_ftell(f) ((size_t)((f)->start_offset + (f)->ptr - (f)->start))

// True if we're at end of file
#define nk_feof(f) ((f)->ptr == (f)->end)


// Set current file position and return character there
// pos must be 0..file_size, inclusive.  It's allowed to set file position to the end of the file, in which case -1 is returned
int nk_fseek_slow(nkinfile_t *f, size_t pos);

// Set current file position and return character there
// Target position must be within current buffer
#define nk_fseek_fast(f, n) (*((f)->ptr = ((n) - (f)->start_offset) + (f)->start))

// True if target is within current buffer
#define nk_contained(f, n) ((n) >= (f)->start_offset && (n) < (f)->start_offset + (f)->len)

// Set current file position and return character there
// n must be 0..file_size.  It's allowed to set file position to the end of the file, in which case -1 is returned
// Automatically uses fast version of fseek if requested position is within current buffer
#define nk_fseek(f, n) (nk_contained((f), (n)) ? nk_fseek_fast((f), (n)) : nk_fseek_slow((f), (n)))

// Relative set current file position and return character there
// n + nk_ftell(f) must be 0..file_size inclusive.
// It's allowed to set file position to the end of the file, in which case -1 is returned
// Automatically uses fast version of fseek if requested position is within current buffer
#define nk_fseek_rel(f, n) nk_fseek((f), nk_ftell(f) + (n))


// Read current character.  Only call if you know you're not already at end of file
//  Always returns valid file character
#define nk_fpeek_fast(f) (*(f)->ptr)

// Same as above, but you may call if you are already at end of file
//  Returns -1 if at end of file
#define nk_fpeek(f) (nk_feof(f) ? -1 : nk_fpeek_fast(f))

// Peek from a specified relative offset without changing file position
#define nk_fpeek_rel(f, n) (nk_contained((f), nk_ftell(f) + (n)) ? *((f)->ptr + (n)) : (((f)->tmp_pos = nk_ftell(f)), (f)->tmp = nk_fseek((f), nk_ftell(f) + (n)), nk_fseek((f), (f)->tmp_pos), (f)->tmp))

// Peek from a specified absolute offset without changing file position
#define nk_fpeek_abs(f, n) (nk_contained((f), (n)) ? *((f)->ptr + (n) - (f)->start_offset) : (((f)->tmp_pos = nk_ftell(f)), (f)->tmp = nk_fseek((f), (n)), nk_fseek((f), (f)->tmp_pos), (f)->tmp))


// Advance, then read character at that position
// Only safe to call if you already know that you're not at EOF to begin with
// Returns -1 if we end up at EOF
#define nk_fnext_fast(f) (++(f)->ptr == (f)->end ? nk_fseek_slow((f), (f)->start_offset + (f)->len) : nk_fpeek_fast(f))

// Advance, then read character at that position
// Returns -1 if we were already at EOF or if we end up at EOF
#define nk_fnext(f) (nk_feof(f) ? -1 : nk_fnext_fast(f))

// Read current character, then advance to next position
// Only safe to call if you know we're not already at end of file
// Always returns valid character
#define nk_fgetc_fast(f) (((f)->tmp = *(f)->ptr++), (nk_feof(f) ? (nk_fseek_slow((f), (f)->len + (f)->start_offset), (f)->tmp) : (f)->tmp))

// Read current character, then advance to next position
// Safe to call even if you start at end of file (returns -1 for this case)
#define nk_fgetc(f) (nk_feof(f) ? -1 : nk_fgetc_fast(f))


// Open a block device
nkinfile_t *nkinfile_open(
    nkinfile_t *f,
    size_t (*block_read)(
        void *block_read_ptr,
        size_t pos,
        unsigned char *buffer,
        size_t block_size
    ),
    void *block_read_ptr,
    size_t block_size,
    unsigned char *buffer
);

// Open a memory block as a file
nkinfile_t *nkinfile_open_mem(nkinfile_t *f, const unsigned char *mem, size_t size);

// Open a string as a file
nkinfile_t *nkinfile_open_string(nkinfile_t *f, const char *s);

// Print rest of file to nkstdout, usually for debugging

int nk_infile_print(nkinfile_t *f);

// Copy rest of infile to outfile

int nk_fcopy(nkoutfile_t *g, nkinfile_t *f);

extern nkinfile_t *nkinfile_null;

#endif
