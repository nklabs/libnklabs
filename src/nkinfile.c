// Copyright 2021 NK Labs, LLC

// Permission is hereby granted, free of charge, to any person obtaining a
// copy of this software and associated documentation files (the
// "Software"), to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to permit
// persons to whom the Software is furnished to do so, subject to the
// following conditions:

// The above copyright notice and this permission notice shall be included
// in all copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. 
// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT
// OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR
// THE USE OR OTHER DEALINGS IN THE SOFTWARE.

#include <string.h>
#include "nkprintf.h"
#include "nkinfile.h"

int nk_fseek_slow(nkinfile_t *f, size_t pos)
{
    size_t offset;

    // Requested position fits in current buffer
    if (pos >= f->start_offset && pos < f->start_offset + f->len)
    {
        f->ptr = f->start + (pos - f->start_offset);
        return *f->ptr;
    }

    // Nope, load necessary block
    offset = (pos & (f->block_size - 1));
    f->start_offset = pos - offset;

    if (f->block_read)
    {
        // Load block.  If end of file is within block, f->len will be less than block_size.
        // If end of file is exactly at start of block, then nothing is loaded and f->len is zero.
        // Otherwise f->len will be equal to block_size.
        f->len = f->block_read(f->block_read_ptr, f->start_offset, f->buffer, f->block_size);
    }
    else
    {
        // We don't do this, because when we have opened a memory block, we want seek to work
        // even after we've reached EOF. 
        // f->len = 0;
    }
    f->end = f->start + f->len;

    if (offset < f->len)
    {
        // Offset is valid
        f->ptr = f->start + offset;
        return *f->ptr;
    }
    else
    {
        // Tried to seek past the end?
        f->ptr = f->start + f->len;
        return -1;
    }
}

nkinfile_t *nkinfile_open(nkinfile_t *f, size_t (*block_read)(void *block_read_ptr, size_t pos, unsigned char *buffer, size_t block_size), void *block_read_ptr, size_t block_size, unsigned char *buffer)
{
    f->buffer = buffer;
    f->start = buffer;
    f->ptr = f->start;
    f->start_offset = 0;
    
    f->block_read_ptr = block_read_ptr;
    f->block_read = block_read;
    f->block_size = block_size;
    f->len = block_size;

    // Read first block into buffer
    if (block_read)
        f->len = block_read(block_read_ptr, f->start_offset, f->buffer, f->block_size);

    f->end = f->start + f->len;

    return f;
}

nkinfile_t *nkinfile_open_mem(nkinfile_t *f, const unsigned char *mem, size_t size)
{
    f->buffer = 0;
    f->start = mem;
    f->ptr = f->start;
    f->start_offset = 0;
    
    f->block_read_ptr = NULL;
    f->block_read = NULL;
    f->block_size = 0;
    f->len = size;

    f->end = f->start + f->len;

    return f;
}

nkinfile_t *nkinfile_open_string(nkinfile_t *f, const char *s)
{
    return nkinfile_open_mem(f, (const unsigned char *)s, strlen(s));
}

int nk_fcopy(nkoutfile_t *g, nkinfile_t *f)
{
    int status = 0;
    while (!nk_feof(f))
        status |= nk_fputc(g, nk_fgetc(f));
    return status;
}

int nk_infile_print(nkinfile_t *f)
{
    return nk_fcopy(nkstdout, f);
}

static nkinfile_t _nkinfile_null =
{
    .buffer = 0,
    .ptr = 0,
    .start = 0,
    .end = 0,
    .len = 0,
    .start_offset = 0,
    .block_read_ptr = 0,
    .block_read = 0,
    .block_size = 0
};

nkinfile_t *nkinfile_null = &_nkinfile_null;
