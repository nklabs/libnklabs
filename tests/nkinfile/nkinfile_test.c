#include <stdio.h>
#include <string.h>
#include "nkinfile.h"

size_t my_block_read_1(void *block_read_ptr, size_t offset, unsigned char *buffer, size_t size)
{
    char *t = (char *)block_read_ptr;
    int l = strlen(t);

    if (offset < l)
    {
        buffer[0] = t[offset];
        return 1;
    }
    else
    {
        return 0;
    }
}

size_t my_block_read_2(void *block_read_ptr, size_t offset, unsigned char *buffer, size_t size)
{
    char *t = (char *)block_read_ptr;
    int l = strlen(t);

    if (offset < l)
    {
        buffer[0] = t[offset];
        buffer[1] = t[offset + 1];
        if (offset + 1 < l)
            return 2;
        else
            return 1;
    }
    else
    {
        return 0;
    }
}

int main(int argc, char *argv[])
{
    unsigned char buf[2];
    nkinfile_t f;
    int c;
    int x;

    printf("Simple string, test nk_fgetc: ");
    nkinfile_open_string(&f, "Hello world!\n");

    while (!nk_feof(&f))
        printf("%c", nk_fgetc(&f));

    printf("Simple string, test nk_fnext: ");

    nk_fseek(&f, 0);

    for (c = nk_fpeek(&f); c != -1; c = nk_fnext(&f))
        printf("%c", c);

    printf("Simple string, test nk_fpeek_abs: ");

    nk_fseek(&f, 0);

    for (x = 0; x != 13; ++x)
        printf("%c", nk_fpeek_abs(&f, x));

    printf("..check eof: %d\n", nk_fpeek_abs(&f, x));
    printf("..check currnt: %c\n", nk_fpeek(&f));

    printf("Simple string, test nk_fpeek_rel: ");

    nk_fseek(&f, 6);

    for (x = -6; x != 7; ++x)
        printf("%c", nk_fpeek_rel(&f, x));

    printf("..check eof: %d\n", nk_fpeek_rel(&f, x));
    printf("..check currnt: %c\n", nk_fpeek(&f));


    printf("Blocks, test nk_fgetc: ");
    nkinfile_open(&f, my_block_read_1, "Hello world!\n", 1, buf);

    while (!nk_feof(&f))
        printf("%c", nk_fgetc(&f));

    printf("Blocks, test nk_fnext: ");

    nk_fseek(&f, 0);

    for (c = nk_fpeek(&f); c != -1; c = nk_fnext(&f))
        printf("%c", c);

    printf("Blocks, test nk_fpeek_abs: ");

    nk_fseek(&f, 0);

    for (x = 0; x != 13; ++x)
        printf("%c", nk_fpeek_abs(&f, x));

    printf("..check eof: %d\n", nk_fpeek_abs(&f, x));
    printf("..check currnt: %c\n", nk_fpeek(&f));

    printf("Blocks, test nk_fpeek_rel: ");

    nk_fseek(&f, 6);

    for (x = -6; x != 7; ++x)
        printf("%c", nk_fpeek_rel(&f, x));

    printf("..check eof: %d\n", nk_fpeek_rel(&f, x));
    printf("..check currnt: %c\n", nk_fpeek(&f));

    printf("Blocks2, test nk_fgetc: ");
    nkinfile_open(&f, my_block_read_2, "Hello world!\n", 1, buf);

    while (!nk_feof(&f))
        printf("%c", nk_fgetc(&f));

    printf("Blocks2, test nk_fnext: ");

    nk_fseek(&f, 0);

    for (c = nk_fpeek(&f); c != -1; c = nk_fnext(&f))
        printf("%c", c);

    printf("Blocks2, test nk_fpeek_abs: ");

    nk_fseek(&f, 0);

    for (x = 0; x != 13; ++x)
        printf("%c", nk_fpeek_abs(&f, x));

    printf("..check eof: %d\n", nk_fpeek_abs(&f, x));
    printf("..check currnt: %c\n", nk_fpeek(&f));

    printf("Blocks2, test nk_fpeek_rel: ");

    nk_fseek(&f, 6);

    for (x = -6; x != 7; ++x)
        printf("%c", nk_fpeek_rel(&f, x));

    printf("..check eof: %d\n", nk_fpeek_rel(&f, x));
    printf("..check currnt: %c\n", nk_fpeek(&f));

    printf("Blocks3, test nk_fgetc: ");
    nkinfile_open(&f, my_block_read_2, "Hello world\n", 1, buf);

    while (!nk_feof(&f))
        printf("%c", nk_fgetc(&f));

    printf("Blocks3, test nk_fnext: ");

    nk_fseek(&f, 0);

    for (c = nk_fpeek(&f); c != -1; c = nk_fnext(&f))
        printf("%c", c);

    printf("Blocks3, test nk_fpeek_abs: ");

    nk_fseek(&f, 0);

    for (x = 0; x != 12; ++x)
        printf("%c", nk_fpeek_abs(&f, x));

    printf("..check eof: %d\n", nk_fpeek_abs(&f, x));
    printf("..check currnt: %c\n", nk_fpeek(&f));

    printf("Blocks3, test nk_fpeek_rel: ");

    nk_fseek(&f, 6);

    for (x = -6; x != 6; ++x)
        printf("%c", nk_fpeek_rel(&f, x));

    printf("..check eof: %d\n", nk_fpeek_rel(&f, x));
    printf("..check currnt: %c\n", nk_fpeek(&f));

    return 0;
}
