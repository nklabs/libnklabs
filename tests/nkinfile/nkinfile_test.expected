Simple string, test nk_fgetc: Hello world!
Simple string, test nk_fnext: Hello world!
Simple string, test nk_fpeek_abs: Hello world!
..check eof: -1
..check currnt: H
Simple string, test nk_fpeek_rel: Hello world!
..check eof: -1
..check currnt: w
Blocks, test nk_fgetc: Hello world!
Blocks, test nk_fnext: Hello world!
Blocks, test nk_fpeek_abs: Hello world!
..check eof: -1
..check currnt: H
Blocks, test nk_fpeek_rel: Hello world!
..check eof: -1
..check currnt: w
Blocks2, test nk_fgetc: Hello world!
Blocks2, test nk_fnext: Hello world!
Blocks2, test nk_fpeek_abs: Hello world!
..check eof: -1
..check currnt: H
Blocks2, test nk_fpeek_rel: Hello world!
..check eof: -1
..check currnt: w
Blocks3, test nk_fgetc: Hello world
Blocks3, test nk_fnext: Hello world
Blocks3, test nk_fpeek_abs: Hello world
..check eof: -1
..check currnt: H
Blocks3, test nk_fpeek_rel: Hello world
..check eof: -1
..check currnt: w
